modules/apps/frontend-js/frontend-js-bundle-config-extender/src/main/java/com/liferay/frontend/js/bundle/config/extender/internal/JSBundleConfigRegistryUtil.java

```diff
@@ -0,0 +1,140 @@
+
+import javax.servlet.ServletContext;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+
+/**
+ * @author Carlos Sierra Andrés
+ */
+public class JSBundleConfigRegistryUtil {
+
+	public static Collection<JSConfig> getJSConfigs() {
+		return _jsConfigs.values();
+	}
+
+	public static long getLastModified() {
+		return _lastModified;
+	}
+
+	public static class JSConfig {
+
+		public ServletContext getServletContext() {
+			return _servletContext;
+		}
+
+		public URL getURL() {
+			return _url;
+		}
+
+		private JSConfig(ServletContext servletContext, URL url) {
+			_servletContext = servletContext;
+			_url = url;
+		}
+
+		private final ServletContext _servletContext;
+		private final URL _url;
+
+	}
+
+	private static final BundleContext _bundleContext;
+	private static final Map<ServiceReference<ServletContext>, JSConfig>
+		_jsConfigs = new ConcurrentSkipListMap<>();
+	private static volatile long _lastModified = System.currentTimeMillis();
+	private static final ServiceTracker
+		<ServletContext, ServiceReference<ServletContext>> _serviceTracker;
+
+	private static class JSBundleConfigServiceTrackerCustomizer
+		implements ServiceTrackerCustomizer
+			<ServletContext, ServiceReference<ServletContext>> {
+
+		@Override
+		public ServiceReference<ServletContext> addingService(
+			ServiceReference<ServletContext> serviceReference) {
+
+			Bundle bundle = serviceReference.getBundle();
+
+			Dictionary<String, String> headers = bundle.getHeaders(
+				StringPool.BLANK);
+
+			String jsConfig = headers.get("Liferay-JS-Config");
+
+			if (jsConfig != null) {
+				URL url = bundle.getEntry(jsConfig);
+
+				if (url != null) {
+					ServletContext servletContext = _bundleContext.getService(
+						serviceReference);
+
+					_jsConfigs.put(
+						serviceReference, new JSConfig(servletContext, url));
+
+					_lastModified = System.currentTimeMillis();
+
+					return serviceReference;
+				}
+			}
+
+			return null;
+		}
+
+		@Override
+		public void modifiedService(
+			ServiceReference<ServletContext> serviceReference,
+			ServiceReference<ServletContext> trackedServiceReference) {
+
+			removedService(serviceReference, trackedServiceReference);
+
+			addingService(serviceReference);
+		}
+
+		@Override
+		public void removedService(
+			ServiceReference<ServletContext> serviceReference,
+			ServiceReference<ServletContext> trackedServiceReference) {
+
+			JSConfig jsConfig = _jsConfigs.remove(serviceReference);
+
+			if (jsConfig != null) {
+				_bundleContext.ungetService(serviceReference);
+
+				_lastModified = System.currentTimeMillis();
+			}
+		}
+
+	}
+
+	static {
+		Bundle bundle = FrameworkUtil.getBundle(
+			JSBundleConfigRegistryUtil.class);
+
+		_bundleContext = bundle.getBundleContext();
```

Comment:
Hey @MrJohn1911, please remove the `_bundleContext` static field in favor of using an instance field in the  JSBundleConfigServiceTrackerCustomizer.